main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala

```diff
@@ -274,59 +275,80 @@ object TypeReduction {
-
-          case exact :: Nil => JavaMethodResolution.Resolved(exact)
-
-          // Multiple exact matches are impossible in Java.
-          case _ :: _ :: _ =>
-            throw InternalCompilerException("Unexpected multiple exact matches for Java method", loc)
-        }
-    }
-  }
-
-  /**
-    * Removes methods of super-classes that are overridden with the same types.
-    */
-  private def candidateMethods(clazz: Class[?], methodName: String, ts: List[Type], static: Boolean): List[Method] = {
-    // this list contains e.g. both `StringBuilder.appendCodePoint(int)` and `AbstractStringBuilder.appendCodePoint(int)`.
-    val allCandidates = JvmOps.getMethods(clazz).filter(isCandidateMethod(_, methodName, static, ts))
-
-    def notOverridden(method: Method): Boolean = {
-      // this is a very hardcoded hack to make the standard library compile
-      // conceptually we should whether the defining classes are subtypes of eachother with exact signatures
-      // but that requires the Class object, which getMethods don't give us
-      method.getReturnType.equals(Void.TYPE) ||
-        method.getReturnType.isPrimitive ||
-        method.getReturnType.isArray ||
-        java.lang.reflect.Modifier.isInterface(method.getReturnType.getModifiers) ||
-        !(java.lang.reflect.Modifier.isAbstract(method.getReturnType.getModifiers) && !static)
+    val tparams = ts.map(getJavaType)
+    val m = MethodUtils.getMatchingAccessibleMethod(clazz, methodName, tparams *)
+    // We check if we found a method and if its static flag matches.
+    if (m != null && JvmOps.isStatic(m) == static) {
+      // Case 1: We found the method on the clazz.
+      JavaMethodResolution.Resolved(m)
+    } else {
+      // Case 2: We failed to find the method on the clazz.
+      // We make one attempt on java.lang.Object.
+      val classObj = classOf[java.lang.Object]
+      val m = MethodUtils.getMatchingAccessibleMethod(classObj, methodName, tparams *)
+      if (m != null) {
+        // Case 2.1: We found the method on java.lang.Object.
+        JavaMethodResolution.Resolved(m)
+      } else {
+        // Case 2.2: We failed to find the method, so we report an error on the original clazz.
+        JavaMethodResolution.NotFound
+      }
     }
-
-    allCandidates.filter(notOverridden)
   }
 
   /**
-    * Returns `true` if `method` has name `methodName`, is static if `static == true`, and can be
-    * called with arguments types in `ts` according to subtyping.
-    */
-  private def isCandidateMethod(method: Method, methodName: String, static: Boolean, ts: List[Type]): Boolean = {
-    if (method.getName != methodName) return false
-    if (JvmOps.isStatic(method) != static) return false
-    subtypeArguments(method.getParameterTypes, ts)
+   * Returns the Java reflective class object corresponding to the given Flix `tpe`.
+   */
+  private def getJavaType(tpe: Type): Class[?] = tpe match {
```

Comment:
I added my own method here because I want tight control. The old stuff in Resolver is deprecated and will be removed.

Reply:
Ill clean this up later yeah